File: 2018/day3/go/src/main.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"regexp"
	"strconv"
)

func main() {
	file, err := os.Open("../input")
	defer file.Close()
	if err != nil {
		log.Fatal(err)
	}

	var lines []string
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}

	// Parse claims from input to array of maps
	var claims []map[string]int
	for _, line := range lines {
		var pattern = "#(?P<id>\\d+) @ (?P<x>\\d+),(?P<y>\\d+): (?P<xsize>\\d+)x(?P<ysize>\\d+)"
		var re = regexp.MustCompile(pattern)
		match := re.FindStringSubmatch(line)
		var claim = make(map[string]int)
		for i, name := range re.SubexpNames() {
			if i > 0 && i <= len(match) {
				n, err := strconv.Atoi(match[i])
				if err != nil {
					log.Fatal(err)
				}
				claim[name] = n
			}
		}
		claims = append(claims, claim)
	}

	// Create fabric
	const inches = 1000
	fabric := [inches][inches]uint{}

	// Add claims
	for _, claim := range claims {
		for x := claim["x"]; x < claim["x"]+claim["xsize"]; x++ {
			for y := claim["y"]; y < claim["y"]+claim["ysize"]; y++ {
				fabric[x][y]++
			}
		}
	}

	// Sum of multiple claims
	var sumofmultiple int = 0
	for _, x := range fabric {
		for _, y := range x {
			if y >= 2 {
				sumofmultiple++
			}
		}
	}
	fmt.Println(sumofmultiple)

	// test for lonely claims
	for _, claim := range claims {
		var found = true
		for x := claim["x"]; x < claim["x"]+claim["xsize"]; x++ {
			for y := claim["y"]; y < claim["y"]+claim["ysize"]; y++ {
				if fabric[x][y] >= 2 {
					found = false
				}
			}
		}
		if found {
			fmt.Println(claim)
		}
	}
}

File: 2018/day3/python/main.py
import re

inches = 1000
fabric = [[0 for i in range(inches)] for j in range(inches)]

f = open("../input", "r")
fl = f.read().splitlines()

def createClaim(line):
    #           id      x     y      xsize ysize
    regex = r'#(\d+) @ (\d+),(\d+): (\d+)x(\d+)'
    pattern = re.compile(regex)
    m = pattern.match(line)
    claim = { 'id' : int(m.group(1)),
              'x' : int(m.group(2)),
              'y' : int(m.group(3)),
              'xsize' : int(m.group(4)),
              'ysize' : int(m.group(5)) }
    return claim

def addClaim(claim):
    for x in range(claim["xsize"]):
        for y in range(claim["ysize"]):
            fabric[claim["x"]+x][claim["y"]+y] += 1

def testClaim(claim):
    for x in range(claim["xsize"]):
        for y in range(claim["ysize"]):
            if fabric[claim["x"]+x][claim["y"]+y] >= 2:
                return False
    return True

claims = []
for line in fl:
    claims.append(createClaim(line))

for claim in claims:
    addClaim(claim)

sumofmultiple = sum(sum(1 for y in x if y >= 2) for x in fabric)
print(sumofmultiple)

for claim in claims:
    if testClaim(claim):
        print(claim)

